pandora-server-directory/src/networking/connection_client.ts

```diff
@@ -63,12 +72,45 @@ export class ClientConnection extends IncomingConnection<IDirectoryClient, IClie
 			this._account = null;
 		}
 		if (account) {
+			Assert(token?.reason === AccountTokenReason.LOGIN);
 			account.touch();
 			this._account = account;
+			this._loginTokenEventUnsubscribe?.();
+			this._loginToken = token;
+			this._loginTokenEventUnsubscribe = token.onAny((ev) => this.onTokenEvent(ev));
 			this.joinRoom(account.associatedConnections);
 		}
 	}
 
+	private onTokenEvent(event: Partial<TypedEventEmitterEvents<AccountToken>>): void {
+		if (event.tokenDestroyed != null) {
+			Assert(this._loginToken === event.tokenDestroyed);
+			this.onAccountTokenDestroyed();
+		}
+		if (event.extended != null) {
+			Assert(this._loginToken === event.extended);
+			this.sendMessage('loginTokenChanged', {
+				value: event.extended.value,
+				expires: event.extended.expires,
+			});
+		}
+	}
+
+	private onAccountTokenDestroyed(): void {
+		if (this._account != null) {
+			this.setAccount(null);
+			this.sendConnectionStateUpdate();
+		}
+
+		this.character?.markForDisconnect();
```

Comment:
This will never be called - above you already call `setAccount(null)`, which forcefully detaches the character inside.